services/ui_backend_service/data/cache/custom_flowgraph.py

```diff
@@ -210,25 +202,47 @@ def __str__(self):
         return '\n'.join(str(n) for _, n in sorted((n.func_lineno, n)
                                                    for n in self.nodes.values()))
 
-    def output_dot(self):
-
-        def edge_specs():
-            for node in self.nodes.values():
-                for edge in node.out_funcs:
-                    yield '%s -> %s;' % (node.name, edge)
-
-        def node_specs():
-            for node in self.nodes.values():
-                nodetype = 'join' if node.num_args > 1 else node.type
-                yield '"{0.name}"'\
-                      '[ label = <<b>{0.name}</b> | <font point-size="10">{type}</font>> '\
-                      '  fontname = "Helvetica" '\
-                      '  shape = "record" ];'.format(node, type=nodetype)
-
-        return "digraph {0.name} {{\n"\
-               "rankdir=LR;\n"\
-               "{nodes}\n"\
-               "{edges}\n"\
-               "}}".format(self,
-                           nodes='\n'.join(node_specs()),
-                           edges='\n'.join(edge_specs()))
+    def output_steps(self):
```

Comment:
Would be preferable to completely get rid of the custom flowgraph implementation from the backend, and use the client library's FlowGraph  class for parsing a DAG even when only a codepackage is present.

The client flowgraph does not accept a `run.code.flowspec` directly though, and maybe it would be better for the client to simply have a property for `_graph_info` that has clientside fallback logic which uses the codepackage, in case the artifact is missing.